app/com/linkedin/drelephant/spark/heuristics/ConfigurationParametersHeuristic.scala

```diff
@@ -0,0 +1,609 @@
+       .get(SPARK_SQL_SHUFFLE_PARTITIONS).map(_.toInt).getOrElse(SPARK_SQL_SHUFFLE_PARTITIONS_DEFAULT)
+     lazy val sparkExecutorInstances = appConfigurationProperties
+       .get(SPARK_EXECUTOR_INSTANCES).map(_.toInt)
+     lazy val sparkExecutorMemoryOverhead = appConfigurationProperties
+       .get(SPARK_EXECUTOR_MEMORY_OVERHEAD).map(stringToBytes(_))
+     lazy val sparkDriverMemoryOverhead = appConfigurationProperties
+       .get(SPARK_DRIVER_MEMORY_OVERHEAD).map(stringToBytes(_))
+
+     // from observation of user applications, adjusting spark.memory.fraction has not had
+     // much benefit, so always set to the default value.
+     val recommendedMemoryFraction: Double = SPARK_MEMORY_FRACTION_DEFAULT
+
+     // recommended executor configuration values, whose recommended values will be
+     // adjusted as various metrics are analyzed. Initialize to current values.
+     var recommendedNumPartitions: Int = sparkSqlShufflePartitions
+     var recommendedExecutorMemory: Long = sparkExecutorMemory
+     var recommendedExecutorCores: Int = sparkExecutorCores
+
+     // TODO: adjust when there is more information about total container memory usage
+     val recommendedDriverMemoryOverhead: Option[Long] = sparkDriverMemoryOverhead
+
+     private lazy val executorSummaries: Seq[ExecutorSummary] = data.executorSummaries
+     private lazy val driver: ExecutorSummary = executorSummaries.find(_.id == "driver").getOrElse(null)
+
+     if (driver == null) {
+       throw new Exception("No driver found!")
+     }
+
+     val currentParallelism = sparkExecutorInstances.map(_ * sparkExecutorCores)
+
+     val jvmUsedMemoryHeuristic =
+       new JvmUsedMemoryHeuristic(configurationParametersHeuristic.heuristicConfigurationData)
+     val jvmUsedMemoryEvaluator = new JvmUsedMemoryHeuristic.Evaluator(jvmUsedMemoryHeuristic, data)
+
+     val executorGcHeuristic =
+       new ExecutorGcHeuristic(configurationParametersHeuristic.heuristicConfigurationData)
+     val executorGcEvaluator = new ExecutorGcHeuristic.Evaluator(executorGcHeuristic, data)
+     val executorGcSeverity = executorGcEvaluator.severityTimeA
+
+     val stageAnalyzer =
+       new StagesAnalyzer(configurationParametersHeuristic.heuristicConfigurationData, data)
+     val stageAnalysis = stageAnalyzer.getStageAnalysis()
+
+     // check for long running tasks, and increase number of partitions if applicable
+     adjustParametersForLongTasks()
+
+     // check for execution memory spill for any stages, and adjust memory, cores, number of
+     // partitions
+     // (recommendedExecutorCores, recommendedExecutorMemory, recommendedNumPartitions) =
+     adjustParametersForExecutionMemorySpill()
+
+     // check for too much time in GC or OOM, and adjust memory, cores, number of partitions
+     if (hasOOMorGC()) {
+         adjustParametersForGCandOOM()
+     } else {
+       // check if executor memory can be lowered
+         adjustExecutorMemory()
+     }
+
```

Comment:
Whats the rational behind calling first adjustParametersForLongTasks, adjustParametersForExecutionMemorySpill, adjustParametersForGCandOOM . Is there a priority defined to fix long running task and then memory spill

Reply:
The ordering is somewhat arbitrary, but there is a better idea of how much to increase the number of partitions for long tasks, and a reasonable estimate for how much to adjust memory/cores/partitions for execution memory spill. GC and OOM are just a guess. It is going from more exact to less exact.

The adjustment(s) for each could affect the other conditions as well (if more partitions are specified due to long tasks, then this would also help with execution memory spill, and OOM/GC).